# Homework 2: Data Quality Management Systems for AI 

### Scenario
You are a junior data-quality analyst. Your organization plans to train an AI model that predicts whether a transaction is fraudulent. Before the model is trained, you must review the customer transaction data and determine whether it is reliable enough to use.


**Question 1 (3 points):** In one or two sentences, describe what one row represents and name three columns that could be useful to a fraud-detection model.

One row of this data is an individual customer ID represented by an age and state, showing the Transaction date, amount and the amount of transactions in the last 24 hours, as well as the method of channel that this transaction was made. 
The three most helpful for a fraud detection model is, Failed logins, Transactions in the last 24 hours and the transaction account.

In [18]:
import pandas as pd 

customer = pd.read_csv('Data/Lab2_customer_data.csv')
customer.head(50)

,CustomerID,CustomerName,Age,State,ZipCode,TransactionDate,TransactionAmount,TransactionsLast24H,FailedLogins,Channel,DeviceStatus,FraudLabel
0,C1001,John Smith,66.0,Georgia,31401,05/20/25,226.38,4,1,Online,New,0
1,C1002,Maria Garcia,20.0,TN,38103,2025-06-16,131.96,4,4,Online,Trusted,0
2,C1003,Hannah Martinez,47.0,CA,94105,Mar 29 2025,257.38,13,0,Mobile,New,0
3,C1004,Michael Thomas,25.0,tennessee,38103,01/12/25,541.63,2,2,ATM,Trusted,0
4,C1005,Linda Patel,NaN,New York,10001,2025-06-30,121.63,9,2,ATM,Trusted,1
5,C1006,Sophia Taylor,24.0,Georgia,30303,Apr 4 2025,121.63,7,2,Branch,New,0
6,C1007,Grace Wilson,250.0,georgia,31401,02/13/25,568.12,11,5,Online,Trusted,1
7,C1008,Linda Martinez,29.0,NY,11201,2025-01-15,284.95,5,5,Mobile,New,1
8,C1009,Maria Thomas,44.0,NaN,31401,May 8 2025,99.58,4,4,ATM,Trusted,0
9,C1010,Kevin Johnson,35.0,TN,37902,04/20/25,235.37,12,4,ATM,New,1


**Question 2 (2 points):** How many rows and columns are in the raw dataset?

There are 77 rows and 12 columns in this data set

In [19]:
print("Rows:", len(customer))
print("Columns:", len(customer.columns))
print(customer.columns.tolist())

Rows: 77
Columns: 12
['CustomerID', 'CustomerName', 'Age', 'State', 'ZipCode', 'TransactionDate', 'TransactionAmount', 'TransactionsLast24H', 'FailedLogins', 'Channel', 'DeviceStatus', 'FraudLabel']


**Question 3: (6 Points)**  Identify at least six suspicious or inconsistent values. Record the column, the value, and why it may be a problem.

Examples of issue types: a missing value, an impossible age, inconsistent state spelling, inconsistent date format, unusual ZIP code, or duplicate record.

**Issue Types**:
1. Rachel White from Illinois is missing an age
2. Rachel White is using an unknown device
3. Aisha Taylor has a missing zipcode which is inconsistent from Tennessee
4. Aisha Taylor is on an unknown device
5. Rachel White has 12 transactions in the last 24 hours
6. Michael Martinez is an unknown device status and has 13 transactions in the last 24 hours

In [20]:
customer.sample(15, random_state=7)

,CustomerID,CustomerName,Age,State,ZipCode,TransactionDate,TransactionAmount,TransactionsLast24H,FailedLogins,Channel,DeviceStatus,FraudLabel
54,C1055,James Thomas,39.0,Illinois,60601,02/09/25,356.51,3,1,Mobile,Trusted,0
36,C1037,Grace Walker,54.0,tennessee,37902,01/11/25,177.25,3,0,ATM,Trusted,0
59,C1060,Ethan Smith,24.0,TX,75201,Apr 8 2025,340.09,3,3,Mobile,Trusted,0
17,C1018,Rachel White,NaN,Illinois,60601,Jun 16 2025,193.85,12,2,Branch,Unknown,0
41,C1042,Marcus Jackson,62.0,IL,60601,Mar 9 2025,171.69,1,0,ATM,Trusted,1
12,C1013,Linda Smith,62.0,Florida,33130,03/29/25,182.30,11,4,ATM,New,1
27,C1028,Maria Walker,49.0,Illinois,60601,01/18/25,204.25,1,4,Online,New,1
15,C1016,Rachel Thomas,50.0,Tennessee,37201,05/26/25,-125.50,6,2,Mobile,Trusted,0
40,C1041,Sophia Martinez,20.0,Texas,75201,2025-03-13,278.02,7,2,ATM,New,0
2,C1003,Hannah Martinez,47.0,CA,94105,Mar 29 2025,257.38,13,0,Mobile,New,0


**Question 4: (6 Points)** Which columns contain missing values? Which missing field do you believe is most harmful to fraud analysis, and why?

There are missing values in Age, State, ZipCode, Transaction Date, and Transaction Amount, with a total number of missing values being 12. The Transaction amount and the Transaction date are the two most harmful to the transaction analysis. This is potentially harmful because the amount cannot be traced to a number, so it could be thousands of dollars from this, the date is important as well to be able to know when the transaction was made so that users can see when they bought whatever objects they purchased.

In [21]:
missing = customer.isna().sum()
print(missing[missing > 0])
print("Total missing cells:", customer.isna().sum().sum())

Age                  4
State                3
ZipCode              2
TransactionDate      2
TransactionAmount    1
dtype: int64
Total missing cells: 12


**Question 5: (5 Points)** How many exact duplicate rows exist? Explain one way duplicates could distort an AI model or a business report.

There are 5 sets of duplicates. Duplicates could skew the model, as some observations would have more influence over others and more than they should. 

In [22]:
duplicate_count = customer.duplicated().sum()
print("Exact duplicate rows:", duplicate_count)
customer[customer.duplicated(keep=False)].sort_values("CustomerID")

Exact duplicate rows: 5


,CustomerID,CustomerName,Age,State,ZipCode,TransactionDate,TransactionAmount,TransactionsLast24H,FailedLogins,Channel,DeviceStatus,FraudLabel
0,C1001,John Smith,66.0,Georgia,31401,05/20/25,226.38,4,1,Online,New,0
72,C1001,John Smith,66.0,Georgia,31401,05/20/25,226.38,4,1,Online,New,0
7,C1008,Linda Martinez,29.0,NY,11201,2025-01-15,284.95,5,5,Mobile,New,1
73,C1008,Linda Martinez,29.0,NY,11201,2025-01-15,284.95,5,5,Mobile,New,1
18,C1019,Marcus Harris,26.0,Georgia,30303,02/28/25,68.59,2,2,Online,Trusted,0
74,C1019,Marcus Harris,26.0,Georgia,30303,02/28/25,68.59,2,2,Online,Trusted,0
36,C1037,Grace Walker,54.0,tennessee,37902,01/11/25,177.25,3,0,ATM,Trusted,0
75,C1037,Grace Walker,54.0,tennessee,37902,01/11/25,177.25,3,0,ATM,Trusted,0
55,C1056,David White,45.0,N. Carolina,28202,2025-04-18,327.53,8,3,Online,Trusted,0
76,C1056,David White,45.0,N. Carolina,28202,2025-04-18,327.53,8,3,Online,Trusted,0


**Question 6: (5 Points)** List the invalid ages. For each one, explain why it should not be treated as a valid customer age.

The invalid ages are -4, 250, and 0. -4 is an invalid number because -4 is not a realistic age, and cannot buy anything. 0 is too young to buy much of anything. 250 is not currently a known viable age for a person to get to, so that is also an invalid age.

In [23]:
print(customer["Age"].describe())
print(customer.loc[(customer["Age"] < 18) | (customer["Age"] > 100), ["CustomerID", "CustomerName", "Age"]])

count     73.000000
mean      46.945205
std       30.602692
min       -4.000000
25%       29.000000
50%       44.000000
75%       62.000000
max      250.000000
Name: Age, dtype: float64
   CustomerID     CustomerName    Age
6       C1007     Grace Wilson  250.0
33      C1034  Robert Anderson   -4.0
65      C1066     Rachel Patel    0.0


**Question 7 (6 Points)** Group the different values that appear to represent Tennessee. Then identify two other states that use more than one format.

There are a few ways that Tennessee is represented in the data, the main one being Tennessee, followed by TN, tennessee, and Tennesee. Georgia uses a Georgia, GA, and georgia, another state that has multiple formats is California and CA. 

In [24]:
print(customer["State"].value_counts(dropna=False).sort_index())

State
CA                 2
California         3
FL                 2
Florida            4
GA                 3
Georgia           10
IL                 4
Illinois           5
N. Carolina        2
NC                 1
NY                 5
New York           5
North Carolina     1
TN                 4
TX                 5
Tennesee           2
Tennessee          2
Texas              4
georgia            1
new york           3
tennessee          6
NaN                3
Name: count, dtype: int64


**Question 8 (4 points)**: Identify the three date styles used in the dataset. Why can mixed date formats cause problems?

The three date systems are month/day/year, year-month-day, month day year, when running this through automation the AI model would interpret these dates as differently for different reasons and would mix the eventual answers and context.

In [25]:
print(customer["TransactionDate"].head(25).to_string(index=False))

   05/20/25
 2025-06-16
Mar 29 2025
   01/12/25
 2025-06-30
 Apr 4 2025
   02/13/25
 2025-01-15
 May 8 2025
   04/20/25
 2025-01-29
Apr 30 2025
   03/29/25
 2025-02-15
May 16 2025
   05/26/25
 2025-02-12
Jun 16 2025
   02/28/25
 2025-01-19
 May 2 2025
        NaN
 2025-05-18
May 21 2025
   04/13/25


# Data Quality Scorecard

## Completeness

In [26]:
total_cells = customer.shape[0] * customer.shape[1]
filled_cells = customer.notna().sum().sum()
completeness = (filled_cells / total_cells) * 100
print(round(completeness, 1))

98.7


## Duplicate Percentage

In [27]:
duplicate_percentage = (customer.duplicated().sum() / len(customer)) * 100
print(round(duplicate_percentage, 1))

6.5


## Age Validity

In [29]:
valid_age_percentage = (customer["Age"].between(18, 100).sum() / len(customer)) * 100
print(round(valid_age_percentage, 1))

90.9


## Metric Consistency

In [31]:
standard_states = ["Tennessee", "Georgia", "New York", "Illinois",
                   "Texas", "Florida", "California", "North Carolina"]
state_consistency = (customer["State"].isin(standard_states).sum() / customer["State"].notna().sum()) * 100
print(round(state_consistency, 1))

45.9


**Question 9** Complete the scorecard. Which metric shows the greatest weakness? Support your answer with the calculated percentage.

The greatest weakness is the metric consistency, there are a lot of different naming conventions for different states and it should get standardized before na automation of some kind comes along to process the reports. It is only 46% consistent, that is very low and the range that an automation could take all the different values for could make the distribution of data very skewed.

## Data Cleaning

In [32]:
clean = customer.copy()

## Remove Exact Duplicates

In [33]:
clean = clean.drop_duplicates().copy()
print("Rows after duplicate removal:", len(clean))

Rows after duplicate removal: 72


## Standardize Names

In [34]:
state_mapping = {
    "TN":"Tennessee", "Tennesee":"Tennessee", "tennessee":"Tennessee",
    "GA":"Georgia", "georgia":"Georgia",
    "NY":"New York", "new york":"New York",
    "IL":"Illinois", "TX":"Texas", "FL":"Florida",
    "CA":"California", "NC":"North Carolina",
    "N. Carolina":"North Carolina"
}
clean["State"] = clean["State"].replace(state_mapping)
print(clean["State"].value_counts(dropna=False).sort_index())

State
California         5
Florida            6
Georgia           12
Illinois           9
New York          12
North Carolina     3
Tennessee         13
Texas              9
NaN                3
Name: count, dtype: int64


**Question 10** Why is standardizing values usually better than simply deleting every row with an inconsistent state name?

 The data still has meaning, the records are still real records of real people but they mispelled something, that does not mean that their record should get deleted. While deleting may be easier in the short term there are a lot of vulnerabilities that could be overlooked in deleting even one of the rows. 

## Correct Invalid Ages and Missing Ages

**Question 11** What median age was used? Explain one advantage and one limitation of filling missing ages with the median.

The median age was 44. The median is useful because it not really affected much by outliers. The main limitation about this is that many missing ages with the same median value can reduce the natural variation amongst the ages in the data.

In [35]:
clean.loc[~clean["Age"].between(18, 100), "Age"] = pd.NA
median_age = clean["Age"].median()
clean["Age"] = clean["Age"].fillna(median_age)
print("Median age used:", median_age)
print(clean["Age"].describe())

Median age used: 44.0
count    72.000000
mean     45.402778
std      16.830271
min      19.000000
25%      30.500000
50%      44.000000
75%      62.000000
max      75.000000
Name: Age, dtype: float64


In [36]:
median_age

np.float64(44.0)

## Standardize Dates

**Question 12** What date format is now used? What happened to dates that could not be interpreted?

Date format is yyyy-mm-dd. the coerce means taht it will replace the time with NAT which is not a time .

In [37]:
clean["TransactionDate"] = pd.to_datetime(clean["TransactionDate"], errors="coerce")
clean["TransactionDate"] = clean["TransactionDate"].dt.strftime("%Y-%m-%d")
print(clean["TransactionDate"].head(15).to_string(index=False))

C:\Users\dvdsl\AppData\Local\Temp\ipykernel_30692\204228146.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  clean["TransactionDate"] = pd.to_datetime(clean["TransactionDate"], errors="coerce")


2025-05-20
2025-06-16
2025-03-29
2025-01-12
2025-06-30
2025-04-04
2025-02-13
2025-01-15
2025-05-08
2025-04-20
2025-01-29
2025-04-30
2025-03-29
2025-02-15
2025-05-16


## Review Remaining Missing Values

**Question 13** Which missing values remain? Choose one remaining field and recommend whether to fill it, investigate it, or remove the affected row. Explain your decision.

The state and zipcode would be really hard to fill in via median or average guess based on the other states. I think as a data quality perspective none of the rows especially with how few should get removed. I would investigate the zip code or the transaction date to see if we could make a reasonable guess on those or take the average from all the dates what makes the most sense of this transaction at this time. Maybe you could reach out to the user and ask if they know then that would solve some of that as well. The best one that could be fixed is Transaction amount being filled in by average. There is one line where that does not exist so it should not skew the data in any particular direction.

In [38]:
print(clean.isna().sum()[clean.isna().sum() > 0])

State                3
ZipCode              2
TransactionDate      2
TransactionAmount    1
dtype: int64


## Save The Dataset

In [39]:
clean.to_csv("cleaned_customer_data.csv", index=False)
print("File saved.")

File saved.


# Data Enrichment

**NOTES** I ran your code Mr. McFall and it did not work with the difference of types that ZipCode is so I wrote this to fix and overcome that

In [49]:
zip_table = pd.read_csv("Data/Lab2_zip_enrichment.csv")
clean["ZipCode"] = clean["ZipCode"].astype("string").str.zfill(5)

zip_table["ZipCode"] = zip_table["ZipCode"].astype("string").str.zfill(5)

enriched = pd.merge(clean, zip_table, on="ZipCode", how="left")

print(
    enriched[
        ["ZipCode", "County", "Region", "PopulationCategory", "RiskScoreCategory"]
    ].head(12)
)

print(
    enriched[
        ["County", "Region", "PopulationCategory", "RiskScoreCategory"]
    ].isna().sum()
)

   ZipCode         County     Region PopulationCategory RiskScoreCategory
0    31401        Chatham      South              Urban            Medium
1    38103         Shelby      South              Urban              High
2    94105  San Francisco       West              Urban              High
3    38103         Shelby      South              Urban              High
4    10001       New York  Northeast              Urban              High
5    30303         Fulton      South              Urban              High
6    31401        Chatham      South              Urban            Medium
7    11201          Kings  Northeast              Urban              High
8    31401        Chatham      South              Urban            Medium
9    37902           Knox      South              Urban            Medium
10   99999            NaN        NaN                NaN               NaN
11   75201         Dallas      South              Urban              High
County                4
Region        

**Question 14** Which new fields were added? Explain how any two of them could help a fraud-detection model.

The new fields that were added are, County, Region, PopulationCategory and RiskScoreCategory. The risk score category is definitely advantageous for a fraud-detection model. I think county could be helpful narrowing down where purchases are made and if there are oddities within location.

**Question 15** Question 15 (4 points): Some records may not match the ZIP table. Give two possible reasons and explain why an unmatched ZIP should be reviewed.

One main reason that the zipcodes might not match is that these updated zipcodes are more specific without the zipcode. There is also that xipcodes are so near each other that it could be one county but a different zipcode or the other way around.

# Part 6: AI and Cybersecurity Impact

**Question 16** (6 points): Choose two original data-quality problems and explain how each could cause either a false fraud alert or a missed fraud event.


One of the data quality issues is the lack of DeviceStatus and the other is missing information on the amount of transactions in the list. Without knowledge of the device status we cannot triage a situation with potentially concerned we should be with a device that is unknown. One would be more concerned about a $300 transaction more than a $5 transaction.

**Question 17** Could poor-quality data resemble a data-poisoning attack? Explain one similarity and one difference.

Poor data can resemble a data poisoning attack because the model can learn incorrect patterns affecting its accuracy and reliability. An example of this is intent, poor data quality is sometimes caused accidentally by missing values, duplicates, or incorrect entries. This can happen in an intentional manner too, where attackers will deliberately add or modify data to manipulated how the model behaves.

**Question 18**: Recommend three controls for an AI data pipeline. For each control, state which problem it would detect or prevent.

1. Data Validation rules - this is when there is a need to check for missing values, invalid formats/dates, impossible values, and incorrect data types. An affect of this is that the poor-quality of data or any corrupted data will be detected before it can be used.
2. Duplicate Detection - this will identify and remove repeated records, this will prevent biased results or overpresentation caused by the duplicate data
3. Access Controlls and Logs - Limit who can modify or add to the training data and record/document any change. This will help prevent/detect any unauthorized changes and data-poisoning attacks.

**Question 19** Is the raw dataset ready for AI training? Is the cleaned dataset ready? Give a separate yes/no decision for each and justify both decisions with specific evidence.

The Raw dataset is not ready for training. The raw dataset has too many sets of duplicates which will affect which direction the data is skewed, there are missing values and incorrect formats across the dates. This makes it to where training the model on this current raw data will  make it askew and mess with the model that would be made from it. 

The clean data set, still should not be trained on as there are still some errors with it. Like in the Transaction Date, Transaction amount columns those both have some unknowns that may skew the model and affect the results that would come from it. 